keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,570 @@
+
+If any Pod finalizer removal fails in step 3, the controller manager still
+executes step 4 with the Pods that succeeded.
+
+Steps 2 to 4 might be skipped in the scenario where a status update happened
+too recently and the number of uncounted Pods is a small percentage of
+parallelism.
+
+### Deleted Pods
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller treats it the same as any other Pod.
+That is, once it reaches Failed status, the controller accounts for the Pod and
+then removes the finalizer.
+
+This is different from the legacy tracking, where the Job controller does not
+account for deleted Pods. This is a limitation that this KEP also wants to
+solve.
+
+However, if it's the Job controller what deletes the Pod (when parallelism
+is decreased, for example), the controller removes the finalizer before
+deleting it. Thus, these deletions don't count towards the failures.
+   
+### Pod adoption
+
+If a Job with `.status.uncountedTerminatedPods != nil` can adopt a Pod
+(according to the existing adoption criteria), this Pod might not have a
+finalizer.
+
+The job controller adds the finalizer in the same patch request that modifies
+the owner reference.
+
+### Test Plan
+
+- Unit tests:
+  - Job sync with feature gate enabled.
+  - Removal of finalizers when feature gate is disabled.
+  - Tracking of terminating Pods.
+- Integration tests:
+  - Job tracking with feature enabled.
+  - Tracking of terminating Pods.
+  - Transition from feature enabled to disabled and enabled again.
+- E2E test:
+  - Job tracking with feature enabled.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implementation:
+  - Job tracking without lingering Pods
+  - Removal of finalizer when feature gate is disabled.
+- Tests: unit, integration, E2E
+
+#### Alpha -> Beta Graduation
+
+- Support for [Indexed Jobs](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+- Metrics:
+  - latency
+  - errors
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance.
+- Job tracking scales to 10^5 completions.
+
+### Upgrade / Downgrade Strategy
+
+When the feature `JobTrackingWithoutLingeringPods` is enabled for the first
+time, the cluster can have Jobs whose Pods don't have the
+`batch.kubernetes.io/job-completion` finalizer. It would be hard to add the
+finalizer to all Pods while preventing race conditions.
+
+We use `.status.uncountedTerminatedPods != nil` to indicate whether the Job
+was created after the feature was enabled. If this field is nil, the Job
+controller tracks Pods using the legacy tracking.
+
+The kube-apiserver sets `.status.uncountedTerminatedPods` to an empty struct
+when the feature gate `JobTrackingWithoutLingeringPods` is enabled, at Job
+creation. In alpha, apiserver leaves `.status.uncountedTerminatedPods = nil`
+for [Indexed Jobs](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
```

Comment:
This is not needed, it's solemnly  controller responsibility to set .status field.

Reply:
How do you suggest we detect if a Job was existing or not?

Reply:
If the feature-gate is on, the controller will look at that field, in all other cases it can safely ignore it and behave as it does today. Just like you explained that in a different spot where you said that in the worst case where a pod was already removed it will re-create those. In all other cases it will count the already completed ones.

Reply:
Spoke with @alculquicondor on slack, we've agreed that the flag will be for both kcm and kas, kas will be responsible for clearing this field when the feature gate is off.

Reply:
Leaving as discussed offline.